### PR TITLE
Performance improvement for draw functions

### LINKED ISSUE #2517: Refactor `set_at()` and it's usage in draw.c
Currently the function `set_at()` in draw.c looks like this:

```C
static int
set_at(SDL_Surface *surf, int x, int y, Uint32 color)
{
    SDL_PixelFormat *format = surf->format;
    Uint8 *pixels = (Uint8 *)surf->pixels;
    Uint8 *byte_buf, rgb[4];

    if (x < surf->clip_rect.x || x >= surf->clip_rect.x + surf->clip_rect.w ||
        y < surf->clip_rect.y || y >= surf->clip_rect.y + surf->clip_rect.h)
        return 0;

    switch (format->BytesPerPixel) {
        case 1:
            *((Uint8 *)pixels + y * surf->pitch + x) = (Uint8)color;
            break;
        case 2:
            *((Uint16 *)(pixels + y * surf->pitch) + x) = (Uint16)color;
            break;
        case 4:
            *((Uint32 *)(pixels + y * surf->pitch) + x) = color;
            break;
        default: /*case 3:*/
            SDL_GetRGB(color, format, rgb, rgb + 1, rgb + 2);
            byte_buf = (Uint8 *)(pixels + y * surf->pitch) + x * 3;
#if (SDL_BYTEORDER == SDL_LIL_ENDIAN)
            *(byte_buf + (format->Rshift >> 3)) = rgb[0];
            *(byte_buf + (format->Gshift >> 3)) = rgb[1];
            *(byte_buf + (format->Bshift >> 3)) = rgb[2];
#else
            *(byte_buf + 2 - (format->Rshift >> 3)) = rgb[0];
            *(byte_buf + 2 - (format->Gshift >> 3)) = rgb[1];
            *(byte_buf + 2 - (format->Bshift >> 3)) = rgb[2];
#endif
            break;
    }
    return 1;
}
```

There has already been one attempt to improve it by making an 'unsafe' version that doesn't do clip checking. 

This does not go far enough. This function is used all over draw.c inside of while loops that draw lines containing hundreds of pixels, and shapes that may contain many lines. 

For each pixel of these lines we are checking what the bit depth of the surface is. **That could be a thousand checks per polygon for a large polygon - but we only need one check.**

I propose, unless `set_at()` is being called directly by a user to change literally a single pixel, we instead make all our tests for surface bit depth (and ideally shape clipping) happen outside of any per pixel while loops and inside of them we call very specific functions that just set the pixel.

e.g.

```C
// This 'safe' version assumes we have determined the surface bit depth but does clipping per pixel,
// ideally we could clip our draw shapes at a higher level
static int set_at_32bit_safe(Uint8 *pixels, int pitch, int x, int y, Uint32 color, SDL_Rect clip_rect)
{
    if (x < clip_rect.x || x >= clip_rect.x + clip_rect.w ||
        y < clip_rect.y || y >= clip_rect.y + clip_rect.h)
        return 0;

    *((Uint32 *)(pixels + y * pitch) + x) = color;
    return 1;
}
//This 'unsafe' version just sets a 32bit surface pixel to a colour and assumes safe postions (i.e. inside the surface)
//and that we have already determined the needed pixel tpye before calling
static void set_at_32bit_unsafe(Uint8 *pixels, int pitch, int x, int y, Uint32 color)
{
    *((Uint32 *)(pixels + y * pitch) + x) = color;
}
```

In all cases we should also prioritise the 32 bit path - the current set_at() has an 8bit surface depth as the first case, before testing for a 16 bit surface, then a 24bit one before finally testing for a 32bit. These switch cases should be in surface commonality order which is 32bit (4byte) then 24bit (3byte), then 8bit (1byte) and finally the rarely seen 16bit (2 byte) surface.

**To Do list:**

- [ ] Identify all the places where set_at() is eventually used inside of a pixel loop. For a refactorings list.
- [ ] Create simple pixel setter functions like the 32bit onese above for all the normal bit depth surfaces (32bit, 24bit, 16bit, 8bit).
- [ ] copy the surface depth switch case that currently sits inside set_at() to sit outside each of the pixel loops (or higher) so we only test the surface bit depth once per user call to draw.
- [ ] use the new simple pixel setter functions inside the pixel loops now separated cleanly by bit depth. 
- [ ] make sure the tests still pass.
- [ ] check the performance versus current main branch across however many draw functions this problem affects.
- [ ] can any of this be macro'd up to reduce repetition?

### PR DESCRIPTION
**Pull request by MightyJosip**
Fix #2517

This is kinda big refactoring of draw module, function changed are single pixel line thickness, arc, circle, ellipse, triangle, and multi pixel thick rectangle. Most of those functions should see solid performance boost (depends on the function). I didn't change circle quadrant (and corresponding round rect), it needs big refactoring, and right now I don't want that to be roadblocker to this. Also I don't think I could accomplish much with thick line

### CODE REVIEW

**Comment by Starbuck5** [comment]
Hey MightyJosip!

I'm just starting to look at this and it's a lot of lines changed, I think it would be helpful if you explained the rationale behind what you've done. I see you've linked #2517 but it seems like what you've done here is more complex than what MyreMylar pointed out.

You say that functions "should" see a speedup, have you done any performance testing?

**Comment by Starbuck5** [comment]
Ok I'm starting to see the approach now, you've used heavy use of `##` macro syntax to conditionally select functions so whole paths are specific to bit depth.

This is not a strategy I've used before. In the blitter code my strategy was to pass code as a parameter argument into macros.

Before this PR, the draw module is 52.5kb, afterwards it is 86.5 kb.

**Comment by Starbuck5** [comment]
Ok I might be up to speed now. Myre was talking about function pointers in heavy loops on discord (that must've been your previous strategy), so now you've done it so that the code path is duplicate-compiled farther up. Now I look at it and one of my first impressions is that it's excessively duplicated.

Maybe this is what we get to eventually or maybe there is a better way? I'm not sure.

**Comment by itzpr3d4t0r** [comment]
Yeah an actual reproducible program and results would make things easier.

**Comment by MightyJosip** [comment]
> Yeah an actual reproducible program and results would make things easier.

Based on the @MyreMylar test in the issue: https://gist.github.com/MightyJosip/dca7901bb48f053bdb10e0f23244afa1

Tested on current release (2.3.2) and my branch. TLDR of the results

| Function tested            | Newest release | My branch |
|----------------------------|----------------|-----------|
| Draw Small Line            | 0.735          | 0.596     |
| Draw Big Line              | 0.803          | 0.450     |
| Draw Small Arc             | 3.181          | 2.487     |
| Draw Big Arc               | 9.893          | 7.395     |
| Draw Small Thick Circle    | 4.125          | 3.820     |
| Draw Big Thick Circle      | 14.773         | 14.331    |
| Draw Small Circle 1 width  | 1.362          | 1.159     |
| Draw Big Circle 1 width    | 2.485          | 1.440     |
| Draw Small Filled Circle   | 1.222          | 1.220     |
| Draw Big Filled Circle     | 14.795         | 13.881    |
| Draw Small Ellipse         | 1.368          | 1.219     |
| Draw Big Ellipse           | 9.893          | 9.354     |
| Draw Small Filled Ellipse  | 0.613          | 0.588     |
| Draw Big Filled Ellipse    | 9.365          | 9.205     |
| Draw Small Thick Rect      | 0.835          | 0.818     |
| Draw Big Thick Rect        | 8.064          | 8.075     |
| Draw Small Filled Triangle | 0.242          | 0.219     |
| Draw Big Filled Triangle   | 18.950         | 18.270    |

**Comment by itzpr3d4t0r** [comment]
This optimization effort is truly impressive and worth celebrating. However, the number of changes and the effort needed to review them make it quite challenging. Since these are mostly macro compositions, it's hard to follow and review them properly. I'm also worried that future improvements will be difficult.

Could we split this into multiple pull requests? For example, one for lines, one for circles, and so on. This would make it easier and faster to review, test, read, and suggest improvements. Let me know what you think.

**Comment by bilhox** [comment]
> This optimization effort is truly impressive and worth celebrating. However, the number of changes and the effort needed to review them make it quite challenging. Since these are mostly macro compositions, it's hard to follow and review them properly. I'm also worried that future improvements will be difficult.
> 
> Could we split this into multiple pull requests? For example, one for lines, one for circles, and so on. This would make it easier and faster to review, test, read, and suggest improvements. Let me know what you think.

Hello @itzpr3d4t0r , I was also thinking about the same thing when I was reading the changes. I'm also seeing 2 PR (this one and #2080) with the same objective. IMO we should open an issue about general draw performance improvements, and link these 2 PRs to the issue, so if anyone else want to work on it, he can see the work of joking and temmie.

Let me know what you think @MightyJosip and @Temmie3754 .
